libraries/operation-graph/src/OperationExecutionManager.ts

```diff
@@ -41,23 +55,26 @@ export class OperationExecutionManager {
    * Group records are metadata-only entities used for tracking the start and end of a set of related tasks.
    * This is the only extent to which the operation graph is aware of Heft phases.
    */
-  private readonly _groupRecordByName: Map<string, OperationGroupRecord>;
+  private readonly _groupRecordByName: Map<string, OperationGroupRecord<TGroupMetadata>>;
   /**
    * The total number of non-silent operations in the graph.
    * Silent operations are generally used to simplify the construction of the graph.
    */
   private readonly _trackedOperationCount: number;
 
-  public constructor(operations: ReadonlySet<Operation>) {
-    const groupRecordByName: Map<string, OperationGroupRecord> = new Map();
+  public constructor(
+    operations: ReadonlySet<Operation<TOperationMetadata>>,
+    deriveGroupMetadata: (metadata: TOperationMetadata) => TGroupMetadata = () => ({}) as TGroupMetadata
```

Comment:
Probably simpler to have a `ReadonlyMap<string, TGroupMetadata>` since we associate the operations with their groups via strings.

Alternatively, we could update the internals to have the group key be the `TGroupMetadata` value itself instead of a simple string.